swaggerpy/client.py

```diff
@@ -1,94 +1,131 @@
-from swaggerpy.swagger_model import create_model_type, Loader
+from swaggerpy.response import HTTPFuture, post_receive
+from swaggerpy.swagger_model import (
+    Loader,
+    create_model_type,
+    is_file_scheme_uri,
+)
 from swaggerpy.swagger_type import SwaggerTypeCheck
 
 log = logging.getLogger(__name__)
 
-SWAGGER_SPEC_TIMEOUT_S = 300
+SWAGGER_SPEC_CACHE_TTL = 300
 
 
-class CachedClient(object):
-    """A wrapper to client which stores the last updated timestamp and the
-    timeout in secs. when the client expires
+class CacheEntry(object):
+    """An entry in the cache. Each item has it's own ttl.
 
-    :param swagger_client: Core SwaggerClient instance
-    :type swagger_client: :class:`SwaggerClient`
-    :param timeout: timeout in seconds after which the client expires
-    :type timeout: int
+    :param item: the item to cache
+    :param ttl: time-to-live in seconds after which the client expires
+    :type  ttl: int
     """
 
-    def __init__(self, swagger_client, timeout, timestamp=None):
-        self.swagger_client = swagger_client
-        self.timeout = timeout
+    def __init__(self, item, ttl, timestamp=None):
+        self.item = item
+        self.ttl = ttl
         self.timestamp = timestamp or time.time()
 
     def is_stale(self, timestamp=None):
         """Checks if the instance has become stale
-        :return: true/false whether client is now stale
+        :return: True if the cache item is stale, False otherwise
         """
         current_time = timestamp or time.time()
-        return self.timestamp + self.timeout < current_time
+        return self.timestamp + self.ttl < current_time
 
 
-class SwaggerClientFactory(object):
-    """Factory to store swagger clients and refetch the api-docs if the client
+class SwaggerClientCache(object):
+    """Cache to store swagger clients and refetch the api-docs if the client
     becomes stale
     """
 
     def __init__(self):
         self.cache = dict()
 
-    def __call__(self, api_docs_url, *args, **kwargs):
-        """
-        :param api_docs_url: url for swagger api docs used to build the client
-        :type api_docs_url: str
-        :param timeout: (optional) Timeout after which api-docs is stale
-        :return: :class:`CachedClient`
-        """
-        # Construct cache key out of api_docs_url
-        if isinstance(api_docs_url, (str, unicode)):
-            key = api_docs_url
-        else:
-            key = json.dumps(api_docs_url)
+    def __contains__(self, key):
+        return key in self.cache and not self.cache[key].is_stale()
 
-        if (key not in self.cache or self.cache[key].is_stale()):
-            self.cache[key] = self.build_cached_client(
-                api_docs_url, *args, **kwargs
-            )
+    def __call__(self, *args, **kwargs):
+        # timeout is backwards compatible with 0.7
+        ttl = kwargs.pop('ttl', kwargs.pop('timeout', SWAGGER_SPEC_CACHE_TTL))
+        key = repr(args) + repr(kwargs)
```

Comment:
I'm in favor of just breaking compatibility and having less crufty code around unless you're going to go through a formal deprecation process.